File: apps/api/app/services/market_data.py
"""Data.gov.in/Agmarknet integration with the existing null fallback."""

from __future__ import annotations

import os
from datetime import date
from typing import Any
from urllib.parse import quote

import httpx


RESOURCE_ID = "9ef84268-d588-465a-a308-a864a43d0070"


def _number_or(value: Any, fallback: float) -> float:
    try:
        parsed = float(value)
        return parsed if parsed else fallback
    except (TypeError, ValueError):
        return fallback


async def fetch_live_agmarknet_prices(crop_name: str, mandi_name: str) -> list[dict[str, Any]] | None:
    api_key = os.getenv("DATA_GOV_IN_API_KEY") or os.getenv("AGMARKNET_API_KEY")
    if not api_key or api_key == "your_data_gov_in_key":
        return None

    clean_crop = crop_name.split(" ")[0].lower()
    url = (
        f"https://api.data.gov.in/resource/{RESOURCE_ID}?api-key={quote(api_key)}"
        f"&format=json&limit=20&filters[commodity]={quote(clean_crop)}"
    )
    try:
        async with httpx.AsyncClient(timeout=httpx.Timeout(5.0, connect=3.0)) as client:
            response = await client.get(url)
        if not response.is_success:
            return None
        records = response.json().get("records", [])
        if not records:
            return None
        today = date.today().isoformat()
        return [
            {
                "state": record.get("state") or "Maharashtra",
                "district": record.get("district") or "",
                "market": record.get("market") or mandi_name,
                "commodity": record.get("commodity") or crop_name,
                "variety": record.get("variety") or "Local",
                "arrival_date": record.get("arrival_date") or today,
                "min_price": _number_or(record.get("min_price"), 1500),
                "max_price": _number_or(record.get("max_price"), 2500),
                "modal_price": _number_or(record.get("modal_price"), 2000),
            }
            for record in records
        ]
    except (httpx.HTTPError, ValueError, TypeError):
        return None

File: lib/market-data.ts
export interface LiveMandiRecord {
  state: string;
  district: string;
  market: string;
  commodity: string;
  variety: string;
  arrival_date: string;
  min_price: number;
  max_price: number;
  modal_price: number;
}

export async function fetchLiveAgmarknetPrices(
  cropName: string,
  mandiName: string
): Promise<LiveMandiRecord[] | null> {
  const apiKey = process.env.DATA_GOV_IN_API_KEY || process.env.AGMARKNET_API_KEY;

  if (!apiKey || apiKey === 'your_data_gov_in_key') {
    return null;
  }

  try {
    // Government of India Open Data Platform (Agmarknet Daily Prices API Resource ID)
    const resourceId = '9ef84268-d588-465a-a308-a864a43d0070';
    const cleanCrop = cropName.split(' ')[0].toLowerCase();
    const url = `https://api.data.gov.in/resource/${resourceId}?api-key=${apiKey}&format=json&limit=20&filters[commodity]=${encodeURIComponent(
      cleanCrop
    )}`;

    const res = await fetch(url, { next: { revalidate: 3600 } });
    if (res.ok) {
      const json = await res.json();
      const records = json.records || [];
      if (records.length > 0) {
        return records.map((r: any) => ({
          state: r.state || 'Maharashtra',
          district: r.district || '',
          market: r.market || mandiName,
          commodity: r.commodity || cropName,
          variety: r.variety || 'Local',
          arrival_date: r.arrival_date || new Date().toISOString().split('T')[0],
          min_price: parseFloat(r.min_price) || 1500,
          max_price: parseFloat(r.max_price) || 2500,
          modal_price: parseFloat(r.modal_price) || 2000,
        }));
      }
    }
  } catch (err) {
    console.warn('Data.gov.in Agmarknet fetch error, falling back to SQLite database:', err);
  }

  return null;
}
